fastlane/actions/analyze_commits_canary.rb

```diff
@@ -0,0 +1,399 @@
+        Actions.lane_context[SharedValues::RELEASE_NEXT_VERSION] = next_version
+        Actions.lane_context[SharedValues::RELEASE_NEXT_PRERELEASE_VERSION] = "#{next_preid}.#{next_pre_ver}"
+
+        success_message = "Next version (#{next_version}) is higher than last version (#{version}). This version should be released."
+        UI.success(success_message) if is_next_version_releasable
+
+        is_next_version_releasable
+      end
+
+      def self.is_codepush_friendly(params)
+        # Begining of the branch is taken for codepush analysis
+        hash = Actions.sh('git rev-list --max-parents=0 HEAD', log: false).chomp
+        next_major = 0
+        next_minor = 0
+        next_patch = 0
+        last_incompatible_codepush_version = '0.0.0'
+
+        # Get commits log between last version and head
+        splitted = get_commits_from_hash(hash: hash)
+        releases = params[:releases]
+        codepush_friendly = params[:codepush_friendly]
+
+        splitted.each do |line|
+          # conventional commits are in format
+          # type: subject (fix: app crash - for example)
+          commit = Helper::SemanticReleaseHelper.parse_commit(
+            commit_subject: line.split("|")[0],
+            commit_body: line.split("|")[1],
+            releases: releases,
+            codepush_friendly: codepush_friendly
+          )
+
+          if commit[:release] == "major" || commit[:is_breaking_change]
+            next_major += 1
+            next_minor = 0
+            next_patch = 0
+          elsif commit[:release] == "minor"
+            next_minor += 1
+            next_patch = 0
+          elsif commit[:release] == "patch"
+            next_patch += 1
+          end
+
+          unless commit[:is_codepush_friendly]
+            last_incompatible_codepush_version = "#{next_major}.#{next_minor}.#{next_patch}"
+          end
+        end
+
+        Actions.lane_context[SharedValues::RELEASE_LAST_INCOMPATIBLE_CODEPUSH_VERSION] = last_incompatible_codepush_version
+      end
+
+      def self.run(params)
+        if params[:canary]
+          is_next_version_releasable = is_releasable_canary(params)
+          return is_next_version_releasable
+        end
+          
+        is_next_version_releasable = is_releasable(params)
+        is_codepush_friendly(params)
+        is_next_version_releasable
+      end
+
+      #####################################################
+      # @!group Documentation
+      #####################################################
+
+      def self.description
+        "Finds a tag of last release and determinates version of next release"
+      end
+
+      def self.details
+        "This action will find a last release tag and analyze all commits since the tag. It uses conventional commits. Every time when commit is marked as fix or feat it will increase patch or minor number (you can setup this default behaviour). After all it will suggest if the version should be released or not."
```

Comment:
nit: consider one of the multiline string syntaxes (e.g., `%q` or heredoc) to make this easier to read without lots of horizontal scrolling